features/events/traps/src/main/java/org/opennms/netmgt/trapd/EventCreator.java

```diff
@@ -54,17 +54,22 @@ public EventCreator(InterfaceToNodeCache cache, EventConfDao eventConfDao) {
         this.eventConfDao = Objects.requireNonNull(eventConfDao);
     }
 
-    public Event createEventFrom(final TrapDTO trapDTO, final String systemId, final String location, final InetAddress trapAddress) {
+    public Event createEventFrom(final TrapDTO trapDTO, final String systemId, final String location, final InetAddress trapAddress,Boolean shouldUseAddressFromVarbind) {
         LOG.debug("{} trap - trapInterface: {}", trapDTO.getVersion(), trapDTO.getAgentAddress());
 
         // Set event data
+        final InetAddress sourceTrapAddress = Optional.ofNullable(shouldUseAddressFromVarbind)
```

Comment:
I think we can simplify this by using trapAddress from TrapDTO when it is valid and fallback on trapAddress from TrapLogDTO when it is not valid.